include/LJ.h

```diff
@@ -78,291 +82,397 @@ class LJD // Late Joiner Data basic class
-        const std::string& desc,
-        bool someone = true,
-        bool show_liveliness = false)
-        : LJD(tp)
+
+    DelayedSnapshot(
+            const std::chrono::steady_clock::time_point tp,
+            const std::string& desc,
+            bool someone = true,
+            bool show_liveliness = false)
+        : LateJoinerData(tp)
         , description(desc)
         , if_someone(someone)
         , show_liveliness_(show_liveliness)
     {
     }
 
-    ~DS() override {}
-
-    DS() = delete;
-    DS(const DS&) = default;
-    DS(DS&&) = default;
-    DS& operator=(const DS&) = default;
-    DS& operator=(DS&&) = default;
+    ~DelayedSnapshot() override
+    {
+    }
 
-    void operator()(DSManager &) override;
+    DelayedSnapshot() = delete;
+    DelayedSnapshot(
+            const DelayedSnapshot&) = default;
+    DelayedSnapshot(
+            DelayedSnapshot&&) = default;
+    DelayedSnapshot& operator =(
+            const DelayedSnapshot&) = default;
+    DelayedSnapshot& operator =(
+            DelayedSnapshot&&) = default;
+
+    void operator ()(
+            DSManager&) override;
 };
 
 // delayed construction of a new subscriber or publisher
-template<class PS>
-void DEC<PS>::operator()(
-        DSManager & man) /*override*/
+template<class PublisherSubscriber>
+void DelayedEndpointCreation<PublisherSubscriber>::operator ()(
+        DSManager& man)  /*override*/
 {
     // Retrieve the corresponding participant
-    Participant* part = man.getParticipant(participant_guid);
+    DomainParticipant* part = man.getParticipant(participant_guid);
+
+    DomainEntity* pubsub = nullptr;
+
+    man.getPubSubEntityFromParticipantGuid<PublisherSubscriber>(participant_guid, pubsub);
 
     if (part == nullptr && owner_event != nullptr)
     {
         part = man.getParticipant(owner_event->participant_guid);
     }
 
     if (!part)
-    {   // invalid participant
-        LOG_ERROR(LJD_traits<PS>::endpoint_type << " cannot be created because no participant assign.");
+    {
+        // invalid participant
+        LOG_ERROR(
+            LateJoinerDataTraits<PublisherSubscriber>::endpoint_type <<
+                " cannot be created because no participant assign.");
         return;
     }
 
+    Topic* topic = nullptr;
+
     // First we must register the type in the associated participant
     if (participant_attributes->topic.getTopicName() == "UNDEF")
     {
+
         // fill in default topic
         participant_attributes->topic = DSManager::builtin_defaultTopic;
 
         // assure the participant has default type registered
-        TopicDataType* pT = nullptr;
-        if (!Domain::getRegisteredType(part, participant_attributes->topic.topicDataType.c_str(), &pT))
+        if (part->find_type(participant_attributes->topic.topicDataType.c_str()).empty())
         {
-            Domain::registerType(part, &DSManager::builtin_defaultType);
+
+            TypeSupport ts(new HelloWorldPubSubType());
```

Comment:
Learning to name variables :stuck_out_tongue_closed_eyes: I suggest more clearer names